lib/chef/provider/package/msu.rb

```diff
@@ -0,0 +1,148 @@
+#
+# Author:: Nimisha Sharad (<nimisha.sharad@msystechnologies.com>)
+# Copyright:: Copyright 2015-2016, Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# msu_package leverages cab_package
+# The contents of msu file are extracted, which contains one or more cab files.
+# The extracted cab files are installed using Chef::Resource::Package::CabPackage
+# Reference: https://support.microsoft.com/en-in/kb/934307
+require "chef/provider/package"
+require "chef/resource/msu_package"
+require "chef/mixin/shell_out"
+require "chef/provider/package/cab"
+require "chef/util/path_helper"
+require "chef/mixin/uris"
+require "chef/mixin/checksum"
+
+class Chef
+  class Provider
+    class Package
+      class Msu < Chef::Provider::Package
+        include Chef::Mixin::ShellOut
+        include Chef::Mixin::Uris
+        include Chef::Mixin::Checksum
+
+        provides :msu_package, os: "windows"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::MsuPackage.new(new_resource.name)
+
+          # download file if source is a url
+          download_source_file if uri_scheme?(new_resource.source)
+
+          # temp directory where the contents of msu file get extracted
+          @temp_directory = Dir.mktmpdir("chef")
+          extract_msu_contents(@new_resource.source, @temp_directory)
+          @cab_files = read_cab_files_from_xml(@temp_directory)
+
+          unless @cab_files.empty?
+            # set the package version
+            # Here is a catch: How to set package version in case of multiple cab files?
+            @new_resource.source = @cab_files.first
+            cab_pkg = Chef::Provider::Package::Cab.new(@new_resource, nil)
+            new_resource.version(cab_pkg.package_version)
```

Comment:
this shouldn't mutate the new_resource since you're mutating the user-intent.

somehow that needs to be state on this provider object (an accessor that gets the right thing in the correct conditions?)